Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.cpp

```diff
@@ -90,4 +93,62 @@ void GraphicsLayerCARemote::moveToContext(RemoteLayerTreeContext& context)
     context.graphicsLayerDidEnterContext(*this);
 }
 
+class GraphicsLayerCARemoteAsyncContentsDisplayDelegate : public GraphicsLayerAsyncContentsDisplayDelegate {
+public:
+    GraphicsLayerCARemoteAsyncContentsDisplayDelegate(IPC::Connection& connection, DrawingAreaIdentifier identifier, WebCore::GraphicsLayer::PlatformLayerID layerID)
+        : m_connection(connection)
+        , m_drawingArea(identifier)
+        , m_layerID(layerID)
+    { }
+
+    bool hasShareableBackend(ImageBuffer& buffer)
+    {
+        auto* backend = buffer.ensureBackendCreated();
+        if (!backend)
+            return false;
+
+        buffer.flushDrawingContext();
+
+        auto* sharing = backend->toBackendSharing();
+        return is<ImageBufferBackendHandleSharing>(sharing);
+    }
+
+    bool tryCopyToLayer(ImageBuffer& buffer) final
+    {
+        if (!hasShareableBackend(buffer))
+            return false;
+        auto clone = buffer.clone();
+        if (!clone)
+            return false;
+        ASSERT(hasShareableBackend(*clone));
+        auto* backend = clone->ensureBackendCreated();
+        if (!backend)
+            return false;
+
+        clone->flushDrawingContext();
+
+        auto* sharing = backend->toBackendSharing();
+        if (!is<ImageBufferBackendHandleSharing>(sharing))
+            return false;
+
+        auto backendHandle = downcast<ImageBufferBackendHandleSharing>(*sharing).createBackendHandle(SharedMemory::Protection::ReadOnly);
+        m_connection->send(Messages::RemoteLayerTreeDrawingAreaProxy::AsyncSetLayerContents(m_layerID, backendHandle), m_drawingArea.toUInt64());
+        return true;
+    }
+
+private:
+    Ref<IPC::Connection> m_connection;
```

Comment:
Holding onto an `IPC::Connection` makes me a little nervous. Can this just call `WebProcess::singleton().parentProcessConnection()` each time?

Reply:
What makes you nervous about holding it?

My concern was that `WebProcess:terminate` could run concurrently with our worker thread trying to retrieve the connection, and we could end up with a dead object.

With the above code, the main-thread can terminate and disconnect the connection, and the worker will just safely drop the message.

That said, I'm fairly sure that the worker thread should have been shutdown by the time we're terminating the WebProcess, but it's hard to prove.